File: backend/registration-service/app/com/radovan/play/providers/HibernateProvider.java
package com.radovan.play.providers;

import com.google.inject.Provider;
import com.radovan.play.utils.HibernateUtil;
import jakarta.inject.Inject;
import org.hibernate.SessionFactory;
import play.inject.ApplicationLifecycle;

import java.util.concurrent.CompletableFuture;

public class HibernateProvider implements Provider<SessionFactory> {

    private final HibernateUtil hibernateUtil;
    private final SessionFactory sessionFactory;

    @Inject
    public HibernateProvider(ApplicationLifecycle lifecycle) {
        this.hibernateUtil = new HibernateUtil();
        this.sessionFactory = hibernateUtil.getSessionFactory();

        // registrujemo shutdown hook
        lifecycle.addStopHook(() -> {
            hibernateUtil.shutdown();
            return CompletableFuture.completedFuture(null);
        });
    }

    @Override
    public SessionFactory get() {
        return sessionFactory;
    }
}

File: backend/infraction-service/app/com/radovan/play/providers/HibernateProvider.scala
package com.radovan.play.providers

import com.google.inject.Provider
import com.radovan.play.utils.HibernateUtil
import jakarta.inject.Inject
import org.hibernate.SessionFactory
import play.api.inject.ApplicationLifecycle

import scala.concurrent.Future

class HibernateProvider @Inject()(lifecycle: ApplicationLifecycle) extends Provider[SessionFactory] {
  private val hibernateUtil = new HibernateUtil
  private val sessionFactory = hibernateUtil.getSessionFactory

  // registrujemo shutdown hook
  lifecycle.addStopHook { () =>
    Future.successful {
      hibernateUtil.shutdown()
    }
  }

  override def get(): SessionFactory = sessionFactory
}
